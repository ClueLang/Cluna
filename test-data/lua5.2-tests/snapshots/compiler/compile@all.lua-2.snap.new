---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
