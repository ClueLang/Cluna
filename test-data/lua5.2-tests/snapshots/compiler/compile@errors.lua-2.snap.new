---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/errors.lua
---
print("testing errors");
local debug = require("debug")
local mt = getmetatable(_G) || {}
local oldmm = mt.__index
mt.__index = nil
global fn doit(s) {
    local f, msg = load(s)
    if f == nil {
        return msg
    }
    local cond, msg = pcall(f)
    return (!cond) && msg
}
global fn checkmessage(prog, msg) {
    local m = doit(prog)
    assert(string.find(m, msg, 1, true));
}
global fn checksyntax(prog, extra, token, line) {
    local msg = doit(prog)
    if !string.find(token, "^<%a") && !string.find(token, "^char%(") {
        token = "'" .. token .. "'"
    }
    token = string.gsub(token, "(%p)", "%%%1")
    local pt = string.format(`^%%[string ".*"%%]:%d: .- near %s$`, line, token)
    assert(string.find(msg, pt));
    assert(string.find(msg, msg, 1, true));
}
assert(doit("error('hi', 0)") == 'hi');
assert(doit("error()") == nil);
if !_no32 {
    assert(doit("table.unpack({}, 1, n=2^30)"));
}
assert(doit("a=math.sin()"));
assert(!doit("tostring(1)") && doit("tostring()"));
assert(doit("tonumber()"));
assert(doit("repeat until 1; a"));
assert(doit("return;;"));
assert(doit("assert(false)"));
assert(doit("assert(nil)"));
assert(doit("function a (... , ...) end"));
assert(doit("function a (, ...) end"));
assert(doit("local t={}; t = t[#t] + 1"));
checksyntax(`
  local a = {4

`, "'}' expected (to close '{' at line 1)", "<eof>", 3);
checkmessage("a=1; bbbb=2; a=math.sin(3)+bbbb(3)", "global 'bbbb'");
checkmessage("a=1; local a,bbbb=2,3; a = math.sin(1) and bbbb(3)", "local 'bbbb'");
checkmessage("a={}; do local a=1 end a:bbbb(3)", "method 'bbbb'");
checkmessage("local a={}; a.bbbb(3)", "field 'bbbb'");
assert(!string.find(doit("a={13}; local bbbb=1; a[bbbb](3)"), "'bbbb'"));
checkmessage("a={13}; local bbbb=1; a[bbbb](3)", "number");
checkmessage("a=(1)..{}", "a table value");
aaa = nil
checkmessage("aaa.bbb:ddd(9)", "global 'aaa'");
checkmessage("local aaa={bbb=1}; aaa.bbb:ddd(9)", "field 'bbb'");
checkmessage("local aaa={bbb={}}; aaa.bbb:ddd(9)", "method 'ddd'");
checkmessage("local a,b,c; (function () a = b+1 end)()", "upvalue 'b'");
assert(!doit("local aaa={bbb={ddd=next}}; aaa.bbb:ddd(nil)"));
checkmessage("local _ENV = {x={}}; a = a + 1", "global 'a'");
checkmessage("b=1; local aaa='a'; x=aaa+b", "local 'aaa'");
checkmessage("aaa={}; x=3/aaa", "global 'aaa'");
checkmessage("aaa='2'; b=nil;x=aaa*b", "global 'b'");
checkmessage("aaa={}; x=-aaa", "global 'aaa'");
assert(!string.find(doit("aaa={}; x=(aaa or aaa)+(aaa and aaa)"), "'aaa'"));
assert(!string.find(doit("aaa={}; (aaa or aaa)()"), "'aaa'"));
checkmessage("print(print < 10)", "function");
checkmessage("print(print < print)", "two function");
_G.D = debug
checkmessage(`
  -- create light udata
  local x = D.upvalueid(function () return debug end, 1)
  D.setuservalue(x, {})
`, "light userdata");
_G.D = nil
checkmessage("(io.write or print){}", "io.write");
checkmessage("(collectgarbage or print){}", "collectgarbage");
local t = {}
for i = 1, 1000 {
    t[(i)] = "a = x" .. i
}
local s = table.concat(t, "; ")
t = nil
checkmessage(s .. "; a = bbb + 1", "global 'bbb'");
checkmessage("local _ENV=_ENV;" .. s .. "; a = bbb + 1", "global 'bbb'");
checkmessage(s .. "; local t = {}; a = t.bbb + 1", "field 'bbb'");
checkmessage(s .. "; local t = {}; t:bbb()", "method 'bbb'");
checkmessage(`aaa=9
repeat until 3==3
local x=math.sin(math.cos(3))
if math.sin(1) == x then return math.sin(1) end   -- tail call
local a,b = 1, {
  {x='a'..'b'..'c', y='b', z=x},
  {1,2,3,4,5} or 3+3<=3+3,
  3+1>3+1,
  {d = x and aaa[x or y]}}
`, "global 'aaa'");
checkmessage(`
local x,y = {},1
if math.sin(1) == 0 then return 3 end    -- return
x.a()`, "field 'a'");
checkmessage(`
prefix = nil
insert = nil
while 1 do
  local a
  if nil then break end
  insert(prefix, a)
end`, "global 'insert'");
checkmessage(`  -- tail call
  return math.sin("a")
`, "'sin'");
checkmessage(`collectgarbage("nooption")`, "invalid option");
checkmessage(`x = print .. "a"`, "concatenate");
checkmessage("getmetatable(io.stdin).__gc()", "no value");
checkmessage(`
local Var
local function main()
  NoSuchName (function() Var=0 end)
end
main()
`, "global 'NoSuchName'");
print('+');
a = {}
setmetatable(a, {
    __index = string
});
checkmessage("a:sub()", "bad self");
checkmessage("string.sub('a', {})", "#2");
checkmessage("('a'):sub{}", "#1");
checkmessage("table.sort({1,2,3}, table.sort)", "'table.sort'");
checkmessage("string.gsub('s', 's', setmetatable)", "setmetatable'");
global fn f(n) {
    local c = coroutine.create(f)
    local a, b = coroutine.resume(c)
    return b
}
assert(string.find(f(), "C stack overflow"));
checkmessage("coroutine.yield()", "outside a coroutine");
f1 = fn () {
    table.sort({
        1, 
        2, 
        3
    }, coroutine.yield);
}
f = coroutine.wrap(fn () {
    return pcall(f1)
})
assert(string.find(select(2, f()), "yield across"));
idsize = 60 - 1
local fn checksize(source) {
    local _, msg = load("x", source)
    msg = string.match(msg, "^([^:]*):")
    assert(msg::len() <= idsize);
}
for i = 60 - 10, 60 + 10 {
    checksize("@" .. string.rep("x", i));
    checksize(string.rep("x", i - 10));
    checksize("=" .. string.rep("x", i));
}
local fn lineerror(s, l) {
    local err, msg = pcall(load(s))
    local line = string.match(msg, ":(%d+):")
    assert((line && line + 0) == l);
}
lineerror("local a\n for i=1,'a' do \n print(i) \n end", 2);
lineerror("\n local a \n for k,v in 3 \n do \n print(k) \n end", 3);
lineerror("\n\n for k,v in \n 3 \n do \n print(k) \n end", 4);
lineerror("function a.x.y ()\na=a+1\nend", 1);
lineerror("a = \na\n+\n{}", 3);
lineerror("a = \n3\n+\n(\n4\n/\nprint)", 6);
lineerror("a = \nprint\n+\n(\n4\n/\n7)", 3);
lineerror("a\n=\n-\n\nprint\n;", 3);
lineerror(`
a
(
23)
`, 1);
lineerror(`
local a = {x = 13}
a
.
x
(
23
)
`, 2);
lineerror(`
local a = {x = 13}
a
.
x
(
23 + a
)
`, 6);
local p = `
function g() f() end
function f(x) error('a', X) end
g()
`
X = 3
lineerror((p), 3);
X = 0
lineerror((p), nil);
X = 1
lineerror((p), 2);
X = 2
lineerror((p), 1);
if !_soft {
    C = 0
    local l = debug.getinfo(1, "l").currentline
    global fn y() {
        C = C + 1
        y();
    }
    local fn checkstackmessage(m) {
        return (string.find(m, "^.-:%d+: stack overflow"))
    }
    assert(checkstackmessage(doit('y()')));
    print('+');
    assert(checkstackmessage(doit('y()')));
    print('+');
    assert(checkstackmessage(doit('y()')));
    print('+');
    C = 0
    local l1
    local fn g(x) {
        l1 = debug.getinfo(x, "l").currentline
        y();
    }
    local _, stackmsg = xpcall(g, debug.traceback, 1)
    print('+');
    local stack = {}
    for line with string.gmatch(stackmsg, "[^\n]*") {
        local curr = string.match(line, ":(%d+):")
        if curr {
            table.insert(stack, tonumber(curr));
        }
    }
    local i = 1
    while stack[(i)] ~= l1 {
        assert(stack[(i)] == l);
        i = i + 1
    }
    assert(i > 15);
    local res, msg = xpcall(error, error)
    assert(!res && type(msg) == 'string');
    print('+');
    local fn f(x) {
        if x == 0 {
            error('a\n');
        } else {
            local aux = fn () {
                return f(x - 1)
            }
            local a, b = xpcall(aux, aux)
            return a, b
        }
    }
    f(3);
    local fn loop(x, y, z) {
        return 1 + loop(x, y, z)
    }
    local res, msg = xpcall(loop, fn (m) {
        assert(string.find(m, "stack overflow"));
        local res, msg = pcall(loop)
        assert(string.find(msg, "error handling"));
        assert(math.sin(0) == 0);
        return 15
    })
    assert(msg == 15);
    res, msg = pcall(fn () {
        for i = 999900, 1000000, 1 {
            table.unpack({}, 1, i);
        }
    })
    assert(string.find(msg, "too many results"));
}
global fn f() {
    error({
        msg = 'x'
    });
}
res, msg = xpcall(f, fn (r) {
    return {
        msg = r.msg .. 'y'
    }
})
assert(msg.msg == 'xy');
a, b, c = xpcall(string.find, error, "alo", "al")
assert(a && b == 1 && c == 2);
a, b, c = xpcall(string.find, fn (x) {
    return {}
}, true, "al")
assert(!a && type(b) == "table" && c == nil);
print('+');
checksyntax("syntax error", "", "error", 1);
checksyntax("1.000", "", "1.000", 1);
checksyntax("[[a]]", "", "[[a]]", 1);
checksyntax("'aa'", "", "'aa'", 1);
checksyntax("\255a = 1", "", "char(255)", 1);
doit('I = load("a=9+"); a=3');
assert(a == 3 && I == nil);
print('+');
lim = 1000
if _soft {
    lim = 100
}
for i = 1, lim {
    doit('a = ');
    doit('a = 4+nil');
}
local fn testrep(init, rep) {
    local s = "local a; " .. init .. string.rep(rep, 400)
    local a, b = load(s)
    assert(!a && string.find(b, "levels"));
}
testrep("a=", "{");
testrep("a=", "(");
testrep("", "a(");
testrep("", "do ");
testrep("", "while a do ");
testrep("", "if a then else ");
testrep("", "function foo () ");
testrep("a=", "a..");
testrep("a=", "a^");
local s = ("a,")::rep(200) .. "a=nil"
local a, b = load(s)
assert(!a && string.find(b, "levels"));
local lim = 127
local s = "local function fooA ()\n  local "
for j = 1, lim {
    s = s .. "a" .. j .. ", "
}
s = s .. "b,c\n"
s = s .. "local function fooB ()\n  local "
for j = 1, lim {
    s = s .. "b" .. j .. ", "
}
s = s .. "b\n"
s = s .. "function fooC () return b+c"
local c = 1 + 2
for j = 1, lim {
    s = s .. "+a" .. j .. "+b" .. j
    c = c + 2
}
s = s .. "\nend  end end"
local a, b = load(s)
assert(c > 255 && string.find(b, "too many upvalues") && string.find(b, "line 5"));
s = "\nfunction foo ()\n  local "
for j = 1, 300 {
    s = s .. "a" .. j .. ", "
}
s = s .. "b\n"
local a, b = load(s)
assert(string.find(b, "line 2"));
mt.__index = oldmm
print('OK');
