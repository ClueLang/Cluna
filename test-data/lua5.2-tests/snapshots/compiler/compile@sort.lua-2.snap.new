---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
