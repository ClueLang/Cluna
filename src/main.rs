#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
