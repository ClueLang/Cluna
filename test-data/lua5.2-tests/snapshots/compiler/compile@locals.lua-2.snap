---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.2-tests/locals.lua
---
print('testing local variables and environments');
local debug = require("debug")
local fn f(x) {
    x = nil
    return x
}
assert(f(10) == nil);
local fn f() {
    local x
    return x
}
assert(f(10) == nil);
local fn f(x) {
    x = nil
    local y
    return x, y
}
assert(f(10) == nil && select(2, f(20)) == nil);
{
    local i = 10
    {
        local i = 100
        assert(i == 100);
    }
    {
        local i = 1000
        assert(i == 1000);
    }
    assert(i == 10);
    if i != 10 {
        local i = 20
    } else {
        local i = 30
        assert(i == 30);
    }
}
f = nil
local f
x = 1
a = nil
load('local a = {}')();
assert(a == nil);
global fn f(a) {
    local _1, _2, _3, _4, _5
    local _6, _7, _8, _9, _10
    local x = 3
    local b = a
    local c, d = a, b
    if (d == b) {
        local x = 'q'
        x = b
        assert(x == 2);
    } else {
        assert(nil);
    }
    assert(x == 3);
    local f = 10
}
local b = 10
local a
loop  {
    local b
    a, b = 1, 2
    assert(a + 1 == b);
} until a + b == 3
assert(x == 1);
f(2);
assert(type(f) == 'function');
local fn getenv(f) {
    local a, b = debug.getupvalue(f, 1)
    assert(a == '_ENV');
    return b
}
assert(getenv(load("a=3")) == _G);
local c = {}
local f = load("a = 3", nil, nil, c)
assert(getenv(f) == c);
assert(c.a == nil);
f();
assert(c.a == 3);
if not_soft {
    local a
    local p = 4
    for i = 2, 31 {
        for j = -3, 3 {
            assert(load(string.format(`local a=%s;
                                        a=a+%s;
                                        assert(a ==2^%s)`, j, p - j, i)))();
            assert(load(string.format(`local a=%s;
                                        a=a-%s;
                                        assert(a==-2^%s)`, -j, p - j, i)))();
            assert(load(string.format(`local a,b=0,%s;
                                        a=b-%s;
                                        assert(a==-2^%s)`, -j, p - j, i)))();
        }
        p = 2 * p
    }
}
print('+');
if rawget(_G, "querytab") {
    collectgarbage("stop");
    local a = {
        {} = 4, 
        3 = 0, 
        alo = 1, 
        a1234567890123456789012345678901234567890 = 10
    }
    local t = querytab(a)
    for k, _ with pairs(a) {
        a[(k)] = nil
    }
    collectgarbage();
    for i = 0, t - 1 {
        local k = querytab(a, i)
        assert(k == nil || type(k) == 'number' || k == 'alo');
    }
}
assert(_ENV == _G);
{
    local _ENV = (fn (...) {
        return ...
    })(_G, dummy)
    {
        local _ENV = {
            assert = assert
        }
        assert(true);
    }
    mt = {
        _G = _G
    }
    local foo, x
    {
        local _ENV = mt
        global fn foo(x) {
            A = x
            {
                local _ENV = _G
                A = 1000
            }
            return fn (x) {
                return A .. x
            }
        }
    }
    assert(getenv(foo) == mt);
    x = foo('hi')
    assert(mt.A == 'hi' && A == 1000);
    assert(x('*') == mt.A .. '*');
    {
        local _ENV = {
            assert = assert, 
            A = 10
        }
        {
            local _ENV = {
                assert = assert, 
                A = 20
            }
            assert(A == 20);
            x = A
        }
        assert(A == 10 && x == 20);
    }
    assert(x == 20);
    print('OK');
    return 5, f
}
